src/Gaufrette/Adapter/OpenStack.php

```diff
@@ -0,0 +1,351 @@
+<?php
+
+namespace Gaufrette\Adapter;
+
+use Gaufrette\Adapter;
+use Gaufrette\Exception\FileNotFound;
+use Gaufrette\Exception\StorageFailure;
+use Gaufrette\Util;
+use OpenStack\Common\Error\BadResponseError;
+use OpenStack\ObjectStore\v1\Models\Container;
+use OpenStack\ObjectStore\v1\Models\StorageObject;
+use OpenStack\ObjectStore\v1\Service;
+
+/**
+ * OpenStack adapter.
+ *
+ * @author  James Watson <james@sitepulse.org>
+ * @author  Daniel Richter <nexyz9@gmail.com>
+ * @author  Nicolas MURE <nicolas.mure@knplabs.com>
+ *
+ * @see http://docs.os.php-opencloud.com/en/latest/services/object-store/v1/objects.html
+ * @see http://refdocs.os.php-opencloud.com/OpenStack/OpenStack.html
+ */
+class OpenStack implements Adapter,
+                           ChecksumCalculator,
+                           ListKeysAware,
+                           MetadataSupporter,
+                           MimeTypeProvider,
+                           SizeCalculator
+{
+    /**
+     * @var Service
+     */
+    protected $objectStore;
+
+    /**
+     * @var string
+     */
+    protected $containerName;
+
+    /**
+     * @var bool
+     */
+    protected $createContainer;
+
+    /**
+     * @var Container
+     */
+    protected $container;
+
+    /**
+     * @param Service $objectStore
+     * @param string  $containerName   The name of the container
+     * @param bool    $createContainer Whether to create the container if it does not exist
+     */
+    public function __construct(Service $objectStore, $containerName, $createContainer = false)
+    {
+        $this->objectStore = $objectStore;
+        $this->containerName = $containerName;
+        $this->createContainer = $createContainer;
+    }
+
+    /**
+     * Returns an initialized container.
+     *
+     * @throws \RuntimeException
+     *
+     * @return Container
+     */
+    protected function getContainer()
+    {
+        if ($this->container) {
+            return $this->container;
+        }
+
+        try {
+            if ($this->objectStore->containerExists($this->containerName)) {
+                return $this->container = $this->objectStore->getContainer($this->containerName);
+            }
+
+            if (!$this->createContainer) {
+                throw new \RuntimeException(sprintf('Container "%s" does not exist.', $this->containerName));
```

Comment:
Could you replace all `\RuntimeException` with `StorageFailure` exception please?